app/services/authn_context_resolver.rb

```diff
@@ -54,18 +67,96 @@ def no_identity_proofing_vot
     end
   end
 
+  def acr_result
+    @acr_result ||= decorate_acr_result_with_user_context(
+      acr_result_with_sp_defaults,
+    )
+  end
+
   def acr_result_with_sp_defaults
+    @acr_result_with_sp_defaults ||= decorate_acr_result_with_sp_defaults(
+      acr_result_without_sp_defaults,
+    )
+  end
+
+  # @param [Vot::Parser::Result] result Parser result to decorate.
+  # @return [Vot::Parser::Result]
+  def decorate_acr_result_with_sp_defaults(result)
     result_with_sp_aal_defaults(
       result_with_sp_ial_defaults(
-        acr_result_without_sp_defaults,
+        result,
       ),
     )
   end
 
+  # With a given {User}, assert the resolved IAL ACR value and corollary requirements
+  # @param [Vot::Parser::Result] result Parser result to decorate.
+  # @return [Vot::Parser::Result]
+  def decorate_acr_result_with_user_context(result)
+    result_decorator = self.method(:no_identity_proofing_acr_result)
+    if result.biometric_comparison?
+      if result.biometric_comparison_required? || user&.identity_verified_with_biometric_comparison?
+        result_decorator = self.method(:biometrics_proofing_acr_result)
+      elsif user&.identity_verified?
+        result_decorator = self.method(:non_biometric_identity_proofing_acr_result)
+      elsif sp_requires_identity_proofing?
+        result_decorator = self.method(:non_biometric_identity_proofing_acr_result)
+      end
+    elsif result.ialmax?
+      if user&.identity_verified?
+        result_decorator = self.method(:non_biometric_identity_proofing_acr_result)
+      end
+    elsif result.identity_proofing?
+      result_decorator = self.method(:non_biometric_identity_proofing_acr_result)
+    elsif sp_requires_identity_proofing? &&
+          result.component_names.none?(Saml::Idp::Constants::IAL1_AUTHN_CONTEXT_CLASSREF) &&
+          result.component_names.none?(Saml::Idp::Constants::LOA1_AUTHN_CONTEXT_CLASSREF)
+      result_decorator = self.method(:non_biometric_identity_proofing_acr_result)
+    end
+    result_decorator.call(result)
+  end
+
+  def biometrics_proofing_acr_result(result)
+    result.with(
+      asserted_ial_acr: Saml::Idp::Constants::IAL2_BIO_REQUIRED_AUTHN_CONTEXT_CLASSREF,
+      aal2?: true,
+      identity_proofing?: true,
+      biometric_comparison?: true,
+      biometric_comparison_required?: true,
+      two_pieces_of_fair_evidence?: true,
+    )
+  end
+
+  def non_biometric_identity_proofing_acr_result(result)
+    result.with(
+      asserted_ial_acr: use_loa? ?
+                          Saml::Idp::Constants::LOA3_AUTHN_CONTEXT_CLASSREF :
+                          Saml::Idp::Constants::IAL2_AUTHN_CONTEXT_CLASSREF,
+      aal2?: true,
+      identity_proofing?: true,
+      biometric_comparison?: false,
+      biometric_comparison_required?: false,
+      two_pieces_of_fair_evidence?: false,
+    )
+  end
+
+  def no_identity_proofing_acr_result(result)
+    result.with(
+      asserted_ial_acr: use_loa? ?
```

Comment:
the original implementation of `info[:ial]` only returned either `Saml::Idp::Constants::IAL1_AUTHN_CONTEXT_CLASSREF` or `Saml::Idp::Constants::IAL2_AUTHN_CONTEXT_CLASSREF`. we should not be changing that implementation, it may break partners expecting a specific value back